packages/synthchron-tauri/src/components/common/TraceVisualizer.tsx

```diff
@@ -0,0 +1,166 @@
+      <Typography variant='h6' gutterBottom display={'inline'}>
+        Trace
+      </Typography>
+      <Typography
+        variant='subtitle1'
+        display={'inline'}
+        sx={{
+          marginLeft: '8px',
+        }}
+      >
+        (showing {first + last}/{trace.trace.events.length} events)
+      </Typography>
+      {trace.trace.events.slice(0, first).map((event, idx) => (
+        <EventLine idx={idx} event={event} key={idx} />
+      ))}
+      <Box width={'100%'} sx={{ marginBottom: '8px', marginTop: '8px' }}>
+        {skip && (
+          <Divider>
+            <Typography variant='caption' fontWeight={100}>
+              Skipped {trace.trace.events.length - first - last} events
+            </Typography>
+          </Divider>
+        )}
+      </Box>
+      {trace.trace.events
+        .slice(-last, trace.trace.events.length)
+        .map((event, idx) => (
+          <EventLine
+            idx={idx + trace.trace.events.length - last}
+            event={event}
+            key={idx}
+          />
+        ))}
+
+      <LastLine
+        reason={trace.exitReason}
+        acceptingState={trace.acceptingState}
+      />
+    </Paper>
+  )
+}
+
+interface EventLineProps {
+  idx: number
+  event: Event
+}
+
+const EventLine: React.FC<EventLineProps> = ({ idx, event }) => (
+  <Paper
+    sx={{
+      marginTop: '8px',
+      padding: '8px',
+      flexDirection: 'row',
+      display: 'flex',
+      alignItems: 'center',
+      height: '32px',
+    }}
+  >
+    <Typography variant='body2' gutterBottom>
+      {idx + 1}
+    </Typography>
+    <Divider
+      sx={{
+        marginLeft: '8px',
+        marginRight: '8px',
+      }}
+      orientation='vertical'
+    />
+    <Box
+      sx={{
+        display: 'flex',
+        flexDirection: 'row',
+        alignItems: 'center',
+        justifyContent: 'center',
+        // flexGrow: 1,
+      }}
+    >
+      <Chip label={'event: ' + event.name} />
+      {/* {event.name} */}
+      {Object.entries(event.meta).map(([key, value]) => (
+        <Chip
+          key={key}
+          label={key + ': ' + value}
+          sx={{
+            marginLeft: '8px',
+          }}
+        />
+      ))}
+    </Box>
+  </Paper>
+)
+
+interface LastLineProps {
+  reason: TraceTerminationReason
+  acceptingState?: string
+}
+
+const LastLine: React.FC<LastLineProps> = ({ reason, acceptingState }) => (
+  <Paper
+    sx={{
+      marginTop: '8px',
+      padding: '8px',
+      flexDirection: 'row',
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+      backgroundColor: 'rgba(0,0,0,0.1)',
+    }}
+  >
+    <Typography variant='body2'>
+      {reason === 'acceptingStateReached' ? (
+        <>
+          Accepting state reached: <Chip label={acceptingState} />
+        </>
+      ) : reason === 'error' ? (
+        'Error'
+      ) : reason === 'noEnabledActivities' ? (
+        'No enabled activities'
+      ) : (
+        'Max events reached'
```

Comment:
it would be good if the end condition was 'not implemented' such that if new termination reasons are added and one forgets to add the reason here, an error will be shown

Reply:
Great point! Will add